### ISSUE TITLE
no-unsafe-readonly-mutable-assignment failing to identify a ReadonlyShallow correctly

### ISSUE THREAD

**Issue by danielnixon** [danielnixon/eslint-plugin-total-functions#745]
This is working:

```ts
        type Thing = { a: string };

        type OtherThing = { readonly a: string };

        // This is correctly inferred to be ReadonlyShallow        
        type ThingArray = readonly Thing[];
        
        // This is correctly inferred to be ReadonlyDeep
        type OtherThingArray = readonly OtherThing[];

        // ReadonlyDeep
        declare const a: OtherThingArray;

        // Flagged (ReadonlyDeep -> ReadonlyShallow)
        const b: ThingArray = a;
```

But this isn't:

```ts
        type Thing = { a: string };

        type OtherThing = { readonly a: string };

        // ReadonlyDeep
        declare const a: readonly OtherThing[];
        
        // Not flagged (it thinks readonly Thing[] is ReadonlyDeep when it should be ReadonlyShallow)
        const b: readonly Thing[] = a;
```

**Comment by danielnixon** [comment]
Looks like: https://github.com/eslint-functional/eslint-plugin-functional/issues/578#issuecomment-1475391803
